lib/block.ml

```diff
@@ -304,45 +311,108 @@ let check_no_errors = function
+        l;
+    set_variables =
+      List.filter_map (function Label.Set (v, x) -> Some (v, x) | _ -> None) l;
+    unset_variables =
+      List.filter_map (function Label.Unset x -> Some x | _ -> None) l;
+    file_inc = get_label (function File x -> Some x | _ -> None) l;
+  }
+
+let mk_ocaml ~config ~contents ~errors =
+  let kind = "OCaml" in
+  match config with
+  | { file_inc = None; part = None; env; non_det; _ } -> (
+      match guess_ocaml_kind contents with
+      | `Code -> Ok (OCaml { env = Ocaml_env.mk env; non_det; errors })
+      | `Toplevel ->
+          Util.Result.errorf "toplevel syntax is not allowed in OCaml blocks." )
+  | { file_inc = Some _; _ } -> label_not_allowed ~label:"file" ~kind
+  | { part = Some _; _ } -> label_not_allowed ~label:"part" ~kind
+
+let mk_cram ~config ~header ~errors =
+  let kind = "shell" in
+  match config with
+  | { file_inc = None; part = None; env = None; non_det; _ } ->
+      check_no_errors errors >>= fun () ->
+      let language =
+        match header with Some (Header.Shell language) -> language | _ -> `Sh
+      in
+      Ok (Cram { language; non_det })
+  | { file_inc = Some _; _ } -> label_not_allowed ~label:"file" ~kind
+  | { part = Some _; _ } -> label_not_allowed ~label:"part" ~kind
+  | { env = Some _; _ } -> label_not_allowed ~label:"env" ~kind
+
+let mk_toplevel ~config ~contents ~errors =
+  let kind = "toplevel" in
+  match config with
+  | { file_inc = None; part = None; env; non_det; _ } -> (
+      match guess_ocaml_kind contents with
+      | `Code ->
+          Util.Result.errorf "invalid toplevel syntax in toplevel blocks."
+      | `Toplevel ->
+          check_no_errors errors >>= fun () ->
+          Ok (Toplevel { env = Ocaml_env.mk env; non_det }) )
+  | { file_inc = Some _; _ } -> label_not_allowed ~label:"file" ~kind
+  | { part = Some _; _ } -> label_not_allowed ~label:"part" ~kind
+
+let mk_include ~config ~header ~errors =
+  let kind = "include" in
+  match config with
+  | { file_inc = Some file_included; part; non_det = None; env = None; _ } -> (
+      check_no_errors errors >>= fun () ->
+      match header with
+      | Some Header.OCaml ->
+          let file_kind = Fk_ocaml { part_included = part } in
+          Ok (Include { file_included; file_kind })
+      | _ -> (
+          match part with
+          | None ->
+              let file_kind = Fk_other { header } in
+              Ok (Include { file_included; file_kind })
+          | Some _ -> label_not_allowed ~label:"part" ~kind:"non-OCaml include"
+          ) )
+  | { file_inc = None; _ } -> label_required ~label:"file" ~kind
+  | { non_det = Some _; _ } ->
+      label_not_allowed ~label:"non-deterministic" ~kind
+  | { env = Some _; _ } -> label_not_allowed ~label:"env" ~kind
+
+let infer_block ~config ~header ~contents ~errors =
+  match config with
+  | { file_inc = Some file_included; non_det; env; part; _ } -> (
       check_not_set
```

Comment:
would it make sense to use `mk_include` here or is the error reporting different?

Reply:
Same remark for the other cases, i.e. once we have infered the type of block, can't use the corresponding `mk_<block-type>` function?